File: weathergpt-frontend/src/types/weather.ts
import { z } from 'zod';

export type WeatherCondition =
  | 'Clear'
  | 'Sunny'
  | 'Partly Cloudy'
  | 'Cloudy'
  | 'Overcast'
  | 'Mist'
  | 'Fog'
  | 'Light Rain'
  | 'Rain'
  | 'Heavy Rain'
  | 'Thunderstorm'
  | 'Snow'
  | 'Sleet'
  | 'Windy';

export interface LocationCoordinates {
  latitude: number;
  longitude: number;
}

export interface LocationInfo {
  id?: string;
  name: string;
  region?: string;
  country: string;
  latitude: number;
  longitude: number;
  timezone?: string;
  is_favorite?: boolean;
}

export interface CurrentWeatherResponse {
  location: LocationInfo;
  temperature: number;
  feels_like: number;
  condition: WeatherCondition | string;
  description: string;
  humidity: number;
  pressure: number;
  wind_speed: number;
  wind_direction: number | string;
  wind_gust?: number;
  visibility: number;
  uv_index: number;
  air_quality_index?: number;
  air_quality_label?: string;
  cloud_cover?: number;
  dew_point?: number;
  sunrise: string;
  sunset: string;
  icon: string;
  timestamp: string;
  is_day: boolean;
}

export interface HourlyForecastItem {
  time: string; // ISO string or format e.g. "14:00"
  timestamp: number;
  temperature: number;
  feels_like: number;
  condition: WeatherCondition | string;
  precipitation_probability: number;
  precipitation_amount?: number; // in mm
  humidity: number;
  wind_speed: number;
  wind_direction?: number | string;
  uv_index?: number;
  icon: string;
  is_day?: boolean;
}

export interface HourlyForecastResponse {
  location: LocationInfo;
  hourly: HourlyForecastItem[];
}

export interface DailyForecastItem {
  date: string; // YYYY-MM-DD
  day_name: string; // e.g. "Monday"
  temperature_high: number;
  temperature_low: number;
  condition: WeatherCondition | string;
  description?: string;
  precipitation_probability: number;
  precipitation_amount?: number;
  humidity: number;
  wind_speed: number;
  uv_index?: number;
  sunrise?: string;
  sunset?: string;
  icon: string;
}

export interface DailyForecastResponse {
  location: LocationInfo;
  forecast: DailyForecastItem[];
}

export interface WeatherAlertItem {
  id: string;
  severity: 'advisory' | 'watch' | 'warning' | 'emergency';
  title: string;
  headline: string;
  description: string;
  instruction?: string;
  effective: string;
  expires: string;
  source: string;
  urgency?: 'immediate' | 'expected' | 'future';
  areas: string[];
}

export interface WeatherAlertsResponse {
  location: LocationInfo;
  alerts: WeatherAlertItem[];
}

export interface AIInsightData {
  summary: string;
  insights: {
    category: 'rain' | 'temperature' | 'outdoor' | 'attire' | 'health' | 'wind';
    title: string;
    description: string;
    importance: 'high' | 'medium' | 'low';
    iconName?: string;
  }[];
  generated_at: string;
}

// Zod schemas for runtime validation
export const LocationInfoSchema = z.object({
  id: z.string().optional(),
  name: z.string(),
  region: z.string().optional(),
  country: z.string(),
  latitude: z.number(),
  longitude: z.number(),
  timezone: z.string().optional(),
});

export const CurrentWeatherSchema = z.object({
  location: LocationInfoSchema,
  temperature: z.number(),
  feels_like: z.number(),
  condition: z.string(),
  description: z.string().default(''),
  humidity: z.number().default(0),
  pressure: z.number().default(1013),
  wind_speed: z.number().default(0),
  wind_direction: z.union([z.number(), z.string()]).default(0),
  visibility: z.number().default(10),
  uv_index: z.number().default(0),
  sunrise: z.string().default('06:00 AM'),
  sunset: z.string().default('06:00 PM'),
  icon: z.string().default('sun'),
  timestamp: z.string().default(() => new Date().toISOString()),
  is_day: z.boolean().default(true),
});

File: backend/schemas/weather.py
from pydantic import BaseModel
from typing import List, Optional, Union
from .location import LocationInfo


class CurrentWeatherResponse(BaseModel):
    location: LocationInfo
    temperature: float
    feels_like: float
    condition: str
    description: str
    humidity: float
    pressure: float
    wind_speed: float
    wind_direction: Union[float, str]
    wind_gust: Optional[float] = None
    visibility: float
    uv_index: float
    air_quality_index: float
    cloud_cover: Optional[float] = None
    dew_point: Optional[float] = None
    sunrise: str
    sunset: str
    icon: str
    timestamp: str
    is_day: bool


class HourlyForecastItem(BaseModel):
    time: str
    timestamp: int
    temperature: float
    feels_like: float
    condition: str
    precipitation_probability: float
    precipitation_amount: Optional[float] = None
    rain: Optional[float] = None
    showers: Optional[float] = None
    humidity: float
    pressure: Optional[float] = None
    wind_speed: float
    wind_speed_80m: Optional[float] = None
    wind_direction: Optional[Union[float, str]] = None
    wind_gust: Optional[float] = None
    visibility: Optional[float] = None
    cloud_cover: Optional[float] = None
    uv_index: Optional[float] = None
    sunshine_duration_s: Optional[float] = None
    is_day: Optional[bool] = None
    icon: str


class HourlyForecastResponse(BaseModel):
    location: LocationInfo
    hourly: List[HourlyForecastItem]


class DailyForecastItem(BaseModel):
    date: str
    day_name: str
    temperature_high: float
    temperature_low: float
    condition: str
    description: Optional[str] = None
    precipitation_probability: float
    precipitation_amount: Optional[float] = None
    rain_sum: Optional[float] = None
    showers_sum: Optional[float] = None
    snowfall_sum: Optional[float] = None
    precipitation_hours: Optional[float] = None
    wind_speed: float
    wind_gust_max: Optional[float] = None
    wind_direction_dominant: Optional[str] = None
    uv_index: Optional[float] = None
    uv_index_clear_sky: Optional[float] = None
    sunshine_duration_hrs: Optional[float] = None
    daylight_duration_hrs: Optional[float] = None
    shortwave_radiation_sum: Optional[float] = None
    moon_phase: Optional[float] = None
    moonrise: Optional[str] = None
    moonset: Optional[str] = None
    sunrise: Optional[str] = None
    sunset: Optional[str] = None
    icon: str


class DailyForecastResponse(BaseModel):
    location: LocationInfo
    forecast: List[DailyForecastItem]


class WeatherAlertItem(BaseModel):
    id: str
    severity: str
    title: str
    headline: str
    description: str
    instruction: Optional[str] = None
    effective: str
    expires: str
    source: str
    urgency: Optional[str] = None
    areas: List[str]


class WeatherAlertsResponse(BaseModel):
    location: LocationInfo
    alerts: List[WeatherAlertItem]


class AIInsightCategory(BaseModel):
    category: str
    title: str
    description: str
    importance: str
    iconName: Optional[str] = None


class AIInsightData(BaseModel):
    summary: str
    insights: List[AIInsightCategory]
    generated_at: str


class AIChatResponse(BaseModel):
    response: str
    insights: Optional[List[str]] = None
    actionable_advice: Optional[dict] = None
    sources: Optional[List[str]] = None
